cuqi/experimental/mcmc/_cwmh.py

```diff
@@ -0,0 +1,222 @@
+                 initial_point=None, **kwargs):
+        super().__init__(target, proposal=proposal, scale=scale,
+                         initial_point=initial_point, **kwargs)
+
+        # set initial scale
+        self.scale = scale
+
+        # set initial acceptance rate
+        self._acc = [np.ones((self.dim))]
+
+    @property
+    def scale(self):
+        """ Get the scale parameter. """
+        return self._scale
+
+    @scale.setter
+    def scale(self, value):
+        """ Set the scale parameter. """
+        if isinstance(value, Number):
+            self._scale = np.ones(self.dim)*value
+        elif isinstance(value, np.ndarray):
+            self._scale = value
+        self._scale_temp = self._scale.copy()
+
+    def validate_target(self):
+        pass # All targets are valid
+
+    @ProposalBasedSamplerNew.proposal.setter
+    # TODO. Check if we can refactor this.
+    # We can work with a validate_proposal method instead?
+    def proposal(self, value):
+        fail_msg = "Proposal should be either None, "+\
+            f"{cuqi.distribution.Distribution.__class__.__name__} "+\
+            "conditioned only on 'location' and 'scale', lambda function, "+\
+            f"or {cuqi.distribution.Normal.__class__.__name__} conditioned "+\
+            "only on 'mean' and 'std'"
+
+        if value is None:
+            self._proposal = cuqi.distribution.Normal(
+                mean=lambda location: location,
+                std=lambda scale: scale,
+                geometry=self.dim,
+            )
+
+        elif isinstance(value, cuqi.distribution.Distribution) and sorted(
+            value.get_conditioning_variables()
+        ) == ["location", "scale"]:
+            self._proposal = value
+
+        elif isinstance(value, cuqi.distribution.Normal) and sorted(
+            value.get_conditioning_variables()
+        ) == ["mean", "std"]:
+            self._proposal = value(
+                mean=lambda location: location, std=lambda scale: scale
+            )
+
+        elif not isinstance(value, cuqi.distribution.Distribution) and callable(
+            value):
+            self._proposal = value
+
+        else:
+            raise ValueError(fail_msg)
+
+    def step(self):
+        # Initialize x_t which is used to store the current CWMH sample
+        x_t = self.current_point.copy()
+
+        # Initialize x_star which is used to store the proposed sample by
+        # updating the current sample component-by-component
+        x_star = self.current_point.copy()
+
+        # Propose a sample x_all_components from the proposal distribution
+        # for all the components
+        target_eval_t = self.current_target
+        if isinstance(self.proposal,cuqi.distribution.Distribution):
+            x_all_components = self.proposal(
+                location= self.current_point, scale=self.scale).sample()
+        else:
+            x_all_components = self.proposal(self.current_point, self.scale)
```

Comment:
This may be out of scope for this PR, but I wonder if we should get rid of the callable proprosal also? And just use distributions. Could create issue for this.

Reply:
good point, I added it to #373